RavenDB-19036: Cluster debug package errored on all nodes

ArieSLV opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19036/Cluster-debug-package-errored-on-all-nodes#focus=Comments-67-395825.0-0

### Additional description

The issue is that we have a timeout of 60 seconds for the operation to complete (for each node). If it isn't done, the operation is canceled, and we move to the next node.
Eventually, we get empty files with no data at all.
We are going to:
- cancel the debug package gracefully and write what we got so far.
- write the last endpoint that we stopped at.
- allow customizing the timeout for that operation (we will need some studio work for that).

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed.

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio:

**On /admin/settings/debugInfo:** For the entire cluster: custom operation timeout duration in seconds for each node.

ArieSLV committed:
RavenDB-19036: Unit test implemented

ArieSLV committed:
RavenDB-19036

ArieSLV committed:
RavenDB-19036:
- cancel the debug package gracefully and write what we got so far.
- write the last endpoint that we stopped at.
- allow customizing the timeout for that operation.

ArieSLV committed:
RavenDB-19036: useless unit-test removed

grisha-kotler commented on src/Raven.Server/Web/Operations/OperationsServerHandler.cs:
```diff
@@ -24,6 +25,26 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
 
+        [RavenAction("/admin/operations/get-operations-id", "GET", AuthorizationStatus.Operator)]
```
Why is it needed?

ArieSLV replied:
Removed

grisha-kotler commented on src/Raven.Client/Documents/Commands/GetNextServerOperationIdCommand.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Raven.Client.Documents.Commands;
+
+public class GetNextServerOperationIdCommand : RavenCommand<long>
```
internal

ArieSLV replied:
Done

grisha-kotler commented on src/Raven.Client/Documents/Commands/KillServerOperationCommand.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Raven.Client.Documents.Commands;
+
+public class KillServerOperationCommand : RavenCommand
```
internal

ArieSLV replied:
Done.

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -175,10 +165,9 @@ public async Task GetInfoPackage()
-            var token = CreateOperationToken();
-
             var operationId = GetLongQueryString("operationId", false) ?? ServerStore.Operations.GetNextOperationId();
+            var timeout = TimeSpan.FromSeconds(GetIntValueQueryString("timeoutInSec", false) ?? _defaultTimeoutForEachNodeInSec);
```
Not needed here, we are going to use it only for the cluster wide

ArieSLV replied:
Ok. Done.

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -99,62 +100,51 @@ public async Task GetClusterWideInfoPackage()
+                var operationId = GetLongQueryString("operationId", false) ?? ServerStore.Operations.GetNextOperationId();
+                
+                var topology = ServerStore.GetClusterTopology(ctx);
+                var timeoutInSec = GetIntValueQueryString("timeoutInSec", false) / topology.AllNodes.Count;
```
```suggestion
                var timeoutInSecPerNode= GetIntValueQueryString("timeoutInSecPerNode", false) ?? 60;
```

ArieSLV replied:
Done

ml054 commented on src/Raven.Server/Web/Operations/OperationsServerHandler.cs:
```diff
@@ -24,6 +25,26 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+                    foreach (Documents.Operations.Operations.Operation operation in operations)
+                    {
+                        writer.WriteStartObject();
+                        writer.WritePropertyName($"Id");
```
nameof?

ArieSLV replied:
This type of operation was unnecessary.

ml054 replied:
I don't follow

I wonder if it is good to use:
```
 writer.WritePropertyName(nameof(operation.Id));
```

ArieSLV replied:
Sorry, Marcin, I didn't make myself clear.
I understand what you recommended, but I removed this code section entirely from the PR due to irrelevance.

karmeli87 commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -99,62 +100,51 @@ public async Task GetClusterWideInfoPackage()
+                var timeoutInSec = GetIntValueQueryString("timeoutInSec", false) / topology.AllNodes.Count;
+                var token = CreateOperationToken(TimeSpan.FromSeconds(timeoutInSec ?? _defaultTimeoutForEachNodeInSec * topology.AllNodes.Count));
+
+                await ServerStore.Operations.AddOperation(null, "Created debug package for all cluster nodes", Operations.Operations.OperationType.DebugPackage, async _ =>
```
let's avoid doing an `await` under the `ctx.OpenReadTransaction()`

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -187,81 +176,111 @@ await using (var ms = new MemoryStream())
+        {
+            var rawStreamCommand = new GetRawStreamResultCommand($"/admin/debug/info-package?operationId={operationId}&timeoutInSec={timeoutInSec}");
+            
+            var requestExecutionTask = requestExecutor.ExecuteAsync(rawStreamCommand, context);
```
why not to pass a CT here?

karmeli87 replied:
and run `KillOperation` if it will throw

grisha-kotler replied:
We want to finish the operation gracefully.
If you pass the `cts` here it will throw and we won't be able to get any result.

karmeli87 replied:
Okay, now I understand... 
I think `timeoutInSec + 2 minutes` would be much simpler here.

karmeli87 replied:
Also, would be great to show some progress (single node / cluster-wide), not sure if this should be part of this issue though.

karmeli87 commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -187,81 +176,111 @@ await using (var ms = new MemoryStream())
+
+        private async Task<Stream> GetDebugInfoFromNodeAsync(JsonOperationContext context, RequestExecutor requestExecutor, long operationId, OperationCancelToken token, int? timeoutInSec)
+        {
+            var rawStreamCommand = new GetRawStreamResultCommand($"/admin/debug/info-package?operationId={operationId}&timeoutInSec={timeoutInSec}");
```
`timeoutInSec ?? _defaultTimeoutForEachNodeInSec`

ArieSLV committed:
RavenDB-19036:
- Clean up
- Access modifiers adjustment
- Careful handling with context.OpenReadTransaction()

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -99,62 +99,51 @@ public async Task GetClusterWideInfoPackage()
+                topology = ServerStore.GetClusterTopology(ctx);
 
+            var timeoutInSecPerNode = GetIntValueQueryString("timeoutInSecPerNode", false) ?? 60;
+            var clusterOperationToken = CreateOperationToken(TimeSpan.FromSeconds(timeoutInSecPerNode * topology.AllNodes.Count));
```
No need for the time here since you use it internally

ArieSLV replied:
Done.

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -187,81 +174,110 @@ await using (var ms = new MemoryStream())
+            var rawStreamCommand = new GetRawStreamResultCommand($"/admin/debug/info-package?operationId={operationId}");
+            var requestExecutionTask = requestExecutor.ExecuteAsync(rawStreamCommand, context);
+
+            using (var delayTaskCts = new CancellationTokenSource())
```
Can be removed and use the `token.Token` directly

ArieSLV replied:
Done.

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -370,27 +378,25 @@ await using (var writer = new AsyncBlittableJsonTextWriter(jsonOperationContext,
                             var endpointOutput = await jsonOperationContext.ReadForMemoryAsync(response.Body, $"read/local endpoint/{route.Path}");
                             jsonOperationContext.Write(writer, endpointOutput);
-                            await writer.FlushAsync();
+                            await writer.FlushAsync(token);
```
What happens if it throws here?

ArieSLV replied:
It will be caught and thrown inside of `WriteForServerOrDatabase` method.
And there will be written this exception as a zip entry.

ArieSLV committed:
RavenDB-19036: Just a little more cleanliness

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -187,81 +174,107 @@ await using (var ms = new MemoryStream())
+                }
+                else
+                {
+                    token.Cancel();
```
ping me about this tomorrow

ArieSLV replied:
You're right. We shouldn't cancel it here.

ArieSLV committed:
RavenDB-19036: One more cancellation.

grisha-kotler commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -187,81 +173,109 @@ await using (var ms = new MemoryStream())
             return rawStreamCommand.Result;
+
+            async Task KillOperation()
+            {
```
Concurrent usage of a context

ArieSLV replied:
Done.

ArieSLV committed:
RavenDB-19036: Stand-alone context for kill operation

ppekrol commented:
Can you rebase and resolve conflicts?

ArieSLV commented:
> Can you rebase and resolve conflicts?

Only tomorrow :(
